tests/unit/test_any.cpp

```diff
@@ -322,5 +326,120 @@ TEST(any, make_any) {
         std::vector<int> ref{1, 2, 3};
         EXPECT_EQ(ref, *vec);
     }
+}
+
+// Tests below are for `any_ptr` class.
+
+TEST(any_ptr, ctor_and_assign) {
+    using util::any_ptr;
+
+    any_ptr p;
+
+    EXPECT_FALSE(p);
+    EXPECT_FALSE(p.has_value());
+
+    int x;
+    any_ptr q(&x);
+
+    EXPECT_TRUE(q);
+    EXPECT_TRUE(q.has_value());
+
+    p = q;
+
+    EXPECT_TRUE(p);
+    EXPECT_TRUE(p.has_value());
+
+    p = nullptr;
+
+    EXPECT_FALSE(p);
+    EXPECT_FALSE(p.has_value());
+
+    p = &x;
+
+    EXPECT_TRUE(p);
+    EXPECT_TRUE(p.has_value());
+
+    p.reset();
+
+    EXPECT_FALSE(p);
+    EXPECT_FALSE(p.has_value());
+
+    p.reset(&x);
+
+    EXPECT_TRUE(p);
+    EXPECT_TRUE(p.has_value());
+
+    p.reset(nullptr);
+
+    EXPECT_FALSE(p);
+    EXPECT_FALSE(p.has_value());
+
+    p = nullptr;
 
+    EXPECT_FALSE(p);
+    EXPECT_FALSE(p.has_value());
 }
+
+
+TEST(any_ptr, ordering) {
+    using util::any_ptr;
+
+    int x[2];
+    double y;
+
+    any_ptr a(&x[0]);
+    any_ptr b(&x[1]);
+
+    EXPECT_LT(a, b);
+    EXPECT_LE(a, b);
+    EXPECT_NE(a, b);
+    EXPECT_GE(b, a);
+    EXPECT_GT(b, a);
+    EXPECT_FALSE(a==b);
+
+    any_ptr c(&y);
+
+    EXPECT_NE(c, a);
```

Comment:
does it make sense to perform comparison of pointers to different types? Certainly, the memory addresses are ordered, however the types are not the same.

Reply:
If the type `any_ptr` is to be ordered for use in e.g `std::map` keys, it must have a well-defined total order for wrapped pointers of different types. In terms of semantics, while one could make a case for `!(a<b || b<a)` for `any_ptr` `a` and `b` with different type information, it would no longer be a total order, as this would not imply `a==b`.

Consequently I think it's best to keep the ordering be the same as the ordering of the addresses of the corresponding objects.